core/state_transition_ovm.go

```diff
@@ -0,0 +1,196 @@
+		msg,
+		msg.From(),
+		&vm.OvmExecutionManager.Address,
+		ret,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return outputmsg, nil
+}
+
+func asOvmMessage(tx *types.Transaction, signer types.Signer) (Message, error) {
+	msg, err := tx.AsMessage(signer)
+	if err != nil {
+		return msg, err
+	}
+
+	if msg.From() == GodAddress {
+		return msg, nil
+	}
+
+	v, r, s := tx.RawSignatureValues()
+	v = new(big.Int).Mod(v, big.NewInt(256))
+	var data = new(bytes.Buffer)
+
+	var sigtype = getSignatureType(msg)
+
+	var target common.Address
+	if tx.To() == nil {
+		target = ZeroAddress
+	} else {
+		target = *tx.To()
+	}
+
+	// Signature type
+	data.WriteByte(byte(sigtype)) // 1 byte: 00 == EOACreate, 01 == EIP 155, 02 == ETH Sign Message
+
+	// Signature data
+	data.Write(v.FillBytes(make([]byte, 1, 1)))   // 1 byte: Signature `v` parameter
+	data.Write(r.FillBytes(make([]byte, 32, 32))) // 32 bytes: Signature `r` parameter
+	data.Write(s.FillBytes(make([]byte, 32, 32))) // 32 bytes: Signature `s` parameter
+
+	if sigtype == 0 {
+		// EOACreate: Encode the transaction hash.
+		data.Write(signer.Hash(tx).Bytes()) // 32 bytes: Transaction hash
+	} else {
+		// EIP 155 or ETH Sign Message: Encode the full transaction data.
+		data.Write(big.NewInt(int64(msg.Nonce())).FillBytes(make([]byte, 2, 2))) // 2 bytes: Nonce
+		data.Write(big.NewInt(int64(msg.Gas())).FillBytes(make([]byte, 3, 3)))   // 3 bytes: Gas limit
+		data.Write(msg.GasPrice().FillBytes(make([]byte, 1, 1)))                 // 1 byte: Gas price
+		data.Write(tx.ChainId().FillBytes(make([]byte, 4, 4)))                   // 4 bytes: Chain ID
+		data.Write(target.Bytes())                                               // 20 bytes: Target address
+		data.Write(msg.Data())                                                   // ?? bytes: Transaction data
+	}
+
+	decompressor := vm.OvmStateDump.Accounts["OVM_SequencerMessageDecompressor"]
+
+	outmsg, err := modMessage(
+		msg,
+		GodAddress,
+		&(decompressor.Address),
+		data.Bytes(),
+	)
+
+	if err != nil {
+		return msg, err
+	}
+
+	return outmsg, nil
+}
+
+func EncodeFakeMessage(
+	msg Message,
+) (Message, error) {
+	var input = []interface{}{
+		big.NewInt(int64(msg.Gas())),
+		msg.To(),
+		msg.Data(),
+	}
+
+	var abi = vm.OvmStateDump.Accounts["mockOVM_ECDSAContractAccount"].ABI
+	output, err := abi.Pack("kall", input...)
+	if err != nil {
+		return nil, err
+	}
+
+	var from = msg.From()
+	return modMessage(
+		msg,
+		from,
+		&from,
+		output,
+	)
+}
+
+func modMessage(
+	msg Message,
+	from common.Address,
+	to *common.Address,
+	data []byte,
+) (Message, error) {
+	queueOrigin, err := getQueueOrigin(msg.QueueOrigin())
+	if err != nil {
+		return nil, err
+	}
+
+	outmsg := types.NewMessage(
+		from,
+		to,
+		msg.Nonce(),
+		msg.Value(),
+		msg.Gas(),
+		msg.GasPrice(),
+		data,
+		false,
+		msg.L1MessageSender(),
+		msg.L1RollupTxId(),
+		queueOrigin,
+		msg.SignatureHashType(),
+	)
+
+	return outmsg, nil
+}
+
+func getSignatureType(
```

Comment:
Only necessary because our sig hash types don't match up with the values in the decompressor contract (or maybe they do now?)

Reply:
I think we updated them, see https://github.com/ethereum-optimism/go-ethereum/blob/b2e6790696a97e5e7c7c0c7f6facd4b66478d51f/core/types/transaction.go#L42